src/main/java/com/ably/kafka/connect/ConfigValueEvaluator.java

```diff
@@ -0,0 +1,43 @@
+package com.ably.kafka.connect;
+
+import org.apache.kafka.connect.sink.SinkRecord;
+
+import java.nio.charset.StandardCharsets;
+
+public class ConfigValueEvaluator {
+
+    public static final String KEY_TOKEN = "${key}";
+    public static final String TOPIC_TOKEN = "${topic}";
+
+    /**
+     * Converts a pattern to a value with the help of given record.
+     * <p>
+     * ${key} will be replaced with record.key()
+     * ${topic} will be replaced with record.topic
+     *
+     * @param record  The SinkRecord to map
+     * @param pattern The pattern to map
+     * @return Evaluated config value given the record and pattern
+     */
+    public String evaluate(SinkRecord record, String pattern) throws IllegalArgumentException{
+        if (pattern == null) {
+            return null;
+        }
+        final byte[] key = (byte[]) record.key();
+        String keyString = null;
+
+        //we only want to evalutate UTF-8 encoded strings
+        if(key != null && ByteArrayUtils.isUTF8Encoded(key)) {
+            keyString = new String(key, StandardCharsets.UTF_8);
+        }
+        if (keyString == null && pattern.contains(KEY_TOKEN)) {
+            throw new IllegalArgumentException("Key is null or not a string type but pattern contains ${key}");
+        }
+
+        if (keyString != null) {
+            return pattern.replace(KEY_TOKEN, keyString).replace(TOPIC_TOKEN, record.topic());
+        } else {
+            return pattern.replace(TOPIC_TOKEN, record.topic());
+        }
+    }
```

Comment:
I'd like to propose to refactor this method to focus its flow around the `pattern` instead of the `key` :wink: Looking at this class tests I figured out that this method's result depends mostly on the provided pattern (given the topic and key values are okay), but I couldn't spot that right away when looking at the implementation :sweat_smile: We could change it to something like the code below :thinking: However, the current code works and I wouldn't like to block the PR so I'll let you decide if that's something we should change :+1: 
```java
    public String evaluate(SinkRecord record, String pattern) throws IllegalArgumentException {
        if (pattern == null) {
            return null;
        } else if (pattern.contains(KEY_TOKEN)) {
            final byte[] key = (byte[]) record.key();

            //we only want to evalutate UTF-8 encoded strings
            if (key != null && ByteArrayUtils.isUTF8Encoded(key)) {
                String keyString = new String(key, StandardCharsets.UTF_8);
                if (pattern.contains(TOPIC_TOKEN)){
                    return pattern.replace(KEY_TOKEN, keyString).replace(TOPIC_TOKEN, record.topic());
                } else {
                    return pattern.replace(KEY_TOKEN, keyString);
                }
            } else {
                throw new IllegalArgumentException("Key is null or not a string type but pattern contains ${key}");
            }
        } else if (pattern.contains(TOPIC_TOKEN)) {
            return pattern.replace(TOPIC_TOKEN, record.topic());
        } else {
            return pattern;
        }
    }
```

Reply:
A good point, but I would argue that current implementation is less verbose and more readable for reasons I will explain below.

Apart from early return, the decision block is this
```
if (keyString != null) {
            return pattern.replace(KEY_TOKEN, keyString).replace(TOPIC_TOKEN, record.topic());
        } else {
            return pattern.replace(TOPIC_TOKEN, record.topic());
        }
```
I prefer early returns for obvious cases where there can be some cases for the function that doesn't require further evaluation or processing (fail fast). And obviously it makes code less less branched. Just to summarise this method's structure
1. Early return based on given argument
2. Key evaluation - input for decision 
3. Decision

I somehow find reading code like this more clean and readable, might be a matter of taste.